jwt.js

```diff
@@ -1,61 +1,114 @@
+
+  let _requestProperty = options.userProperty || options.requestProperty || 'user'
+  let _resultProperty = options.resultProperty
+  const credentialsRequired = typeof options.credentialsRequired === 'undefined' ? true : options.credentialsRequired
 
   fastify.decorate('jwt', {
     sign: sign,
     verify: verify,
-    decode: decode,
-    secret: secret
+    secretCallback: secretCallback
   })
 
   next()
 
-  function sign (payload, options, callback) {
-    assert(payload, 'missing payload')
-    options = options || {}
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
-    }
+  function sign (request, reply, next) {
+    let {secret, resultProperty, ...rest} = options
+    steed.waterfall([
+      function getSecret (callback) {
+        secretCallback({}, {}, callback)
+      },
+      function sign (secret, callback) {
+        JWT.sign(request.body, secret, rest, function (err, token) {
+          if (err) {
+            callback(err)
+          } else {
+            callback(null, token)
+          }
+        })
+      }
+    ], function (err, result) {
+      if (_resultProperty) {
+        reply[_resultProperty] = result
+      } else {
+        request[_requestProperty] = result
+      }
+      return next(err, result)
+    })
+  } // end sign
 
-    if (typeof callback === 'function') {
-      JWT.sign(payload, secret, options, callback)
-    } else {
-      return JWT.sign(payload, secret, options)
-    }
-  }
+  function verify (request, reply, next) {
+    let token
+
+    if (request.headers && request.headers.authorization) {
+      const parts = request.headers.authorization.split(' ')
+      if (parts.length === 2) {
+        const scheme = parts[0]
+        const credentials = parts[1]
 
-  function verify (token, options, callback) {
-    assert(token, 'missing token')
-    assert(secret, 'missing secret')
-    options = options || {}
-    if (typeof options === 'function') {
-      callback = options
-      options = {}
+        if (/^Bearer$/i.test(scheme)) {
+          token = credentials
+        } else {
+          if (credentialsRequired) {
+            return next(new Error('Format is Authorization: Bearer [token]'))
+          } else {
+            return next()
+          }
+        }
+      } else {
+        return next(new Error('Format is Authorization: Bearer [token]'))
+      }
     }
 
-    if (typeof callback === 'function') {
-      JWT.verify(token, secret, options, callback)
-    } else {
-      return JWT.verify(token, secret, options)
+    if (!token) {
+      if (credentialsRequired) {
+        return next(new Error('No authorization token was found'))
+      } else {
+        return next()
+      }
     }
-  }
 
-  function decode (token, options) {
-    assert(token, 'missing token')
-    options = options || {}
-    return JWT.decode(token, options)
-  }
+    let decodedToken = JWT.decode(token, { complete: true }) || {}
```

Comment:
what does this  `complete: true` option do? Can you add a comment explaining? Why are you defaulting this wit `|| {}`.

Reply:
https://github.com/auth0/node-jsonwebtoken#jwtdecodetoken--options

i'll make sure to comment this and mention it in the readme.